### ISSUE TITLE
TODO list

### ISSUE THREAD

**Issue by michael2012z** [rust-vmm/vm-virtio#11]
Hey, maintainers

Can we make a TODO list here for this repo? It will help contributors (like me :) ) know what they can do for the community.

I went through virtio specification (v1.0 and v1.1) and the existing code in the repo. I am trying to make a draft TODO list. Please correct me if I am walking toward wrong direction.

Expecting your comments.

----

TODO LIST:

**Notifications(new feature in spec 1.1):**  
- Implement "Notifications" introduced in spec 1.1. Reference to https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-170003

**New features of Virtqueues (spec 1.1):**
- 2.6.13 Supplying Buffers to The Device (https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-5300013)
- 2.6.14 Receiving Used Buffers From The Device (https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-6000014)

**Packed Virtqueues** (https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-240006). 
- Refer to https://github.com/rust-vmm/vm-virtio/issues/6

**New Device Types (spec 1.1):**
- GPU Device
- Input Device
- Crypto Device
- Socket Device

**Legacy (spec 1.0) device types:**
- Network device
- Block device
- Console device
- Entropy device
- Balloon device
- SCSI device

**Transport options:**
- Virtio over PCI 
- Virtio over MMIO
- Virtio over Channel I/O

**Conformance test?**

**Comment by andreeaflorescu** [comment]
Hey @michael2012z. I am really happy to see people interested in contributing. I think that before talking about specific devices and other features we need to add in Virtio, we first need to agree on the design. The TODO list you created makes a lot of sense, but I believe we should postpone it a bit. The reason is that right now work on this crate is blocked by the PR that introduces an interrupt manager in vm-device. Check this PR: https://github.com/rust-vmm/vm-device/pull/11. Myself, @sameo  and @jiangliu are working on moving forward with the Interrupt controller. There are also a few people working on the general design of Virtio devices. With that said, I don't believe it's a good idea to start the work on other devices right now.

Are you interested in working on other rust-vmm components in the meantime, or is your interest specific to virtio devices?

**Comment by michael2012z** [comment]
Hey, @andreeaflorescu , I am glad to see your reply.

I should introduce myself to the community, but haven't got a good chance. I am Michael Zhao, working in ARM. Now I am trying to contribute to rust-vmm community. @MrXinWang (Xin Wang) is my colleague. Recently we made some PRs in kvm-ioctls repo to begin.

We try to contribute more, so I turned to vm-virtio repo and tried to make a plan for more work. But we are not limited to any specific component. We are also interested in others.

Thank you for introducing the situation of this component. Could you suggest some prioritized work to us?

**Comment by andreeaflorescu** [comment]
> Hey, @andreeaflorescu , I am glad to see your reply.
> 
> I should introduce myself to the community, but haven't got a good chance. I am Michael Zhao, working in ARM. Now I am trying to contribute to rust-vmm community. @MrXinWang (Xin Wang) is my colleague. Recently we made some PRs in kvm-ioctls repo to begin.

I noticed you are very active on kvm-ioctls. Thank you, it is really appreciated!

> Thank you for introducing the situation of this component. Could you suggest some prioritized work to us?

I looked at the list of issues we currently have in [rust-vmm status](https://github.com/orgs/rust-vmm/projects/2). There is not much there, but I can try to come with more issues.

**Comment by stefano-garzarella** [comment]
This project is now moved in https://github.com/rust-vmm/rust-vmm, we are transferring issues to it, but I'm not sure about this one, so I'm closing it. If it's still an issue, please reopen there.
